pkg/apis/metal3/v1alpha1/baremetalhost_types.go

```diff
@@ -147,6 +147,10 @@ type BareMetalHostSpec struct {
 
 	// Description is a human-entered text used to help identify the host
 	Description string `json:"description"`
+
+	// RequestRebootAt requests a reboot after the given timestamp
+	// +optional
+	RequestRebootAt *metav1.Time `json:"requestRebootAt,omitempty"`
```

Comment:
I assume the timestamp here is so you can determine if the reboot was achieved or not. If a reboot was requested yesterday, and another one today, the updated timestamp is used to identify that a new request was made?

It seems awkward, and has the ambiguity @dhellmann identified below around future-dated requests.

Did you consider making a separate CRD to represent a unique reboot request? A small Reconcile function in the BMO, a new "controller", would watch and reconcile the new resource. Each reboot would be initiated by creating one of those resources. Its status would reflect whether the power off and on had taken place or not. In a pinch, the creation timestamp could be compared to the host's `LastBooted` timestamp to determine if the reboot had happened.

Seems like a more explicit way to represent a series of discrete actions. It would also enable history to be preserved.

Example:

```
apiVersion: metal3.io/v1alpha1
kind: BareMetalHostPowerCycle
metadata:
  name: example-pwmdhf
  namespace: default
spec:
  hostReference:
    name: example-host
    namespace: default
status:
  powerOff: <timestamp>
  powerOn: <timestamp>
```

WDYT?

Reply:
> I assume the timestamp here is so you can determine if the reboot was achieved or not. If a reboot was requested yesterday, and another one today, the updated timestamp is used to identify that a new request was made?

I don't really care about that. The timestamp is firstly to ensure that requests from different clients are unique in a way that just setting e.g. `rebootRequest: true` wouldn't be (in that sense it could just as easily be a random UUID or something), and secondly to reduce unnecessary reboots by not initiating any more once the last boot time is later than the last requested reboot.

It is a little awkward, but that's because it's not an interface designed for users, but rather a low-level interface designed to be an implementation detail for some higher-level external CRD like the one you proposed. (Can we use webhooks to limit access to this field to specific classes of CRDs? I'd be fine with that.)

The requirements for fencing are that at some point in time the fencing code decides that a host must be fenced, and it must be able to:

1. stop everything that is running on the host; and
2. determine when everything running on the host has been stopped;
3. regardless of what else might be happening in the system

For this use case the interface is actually quite elegant - the fencing code simply has to copy over the timestamp of the fencing decision and the operator will guarantee the rest.

I'd be interested in being proven wrong, but I don't currently believe that it's possible even in theory to implement those guarantees without having an interface in the BareMetalHost itself that looks pretty similar to this one. All of the other proposed solutions appear to sacrifice correctness in favour of a more general-purpose user-facing interface for sysadmins (something which AFAIK is not actually a requirement, and which in any event could be implemented on top of this as a CRD - and as you point out that is the correct way to implement it anyway). IMHO that isn't a good tradeoff.

Reply:
I would not be too concerned about drawing a line between high-level vs. low-level reboot APIs, or anything "user-facing" vs not. We just need a declarative way to reboot a host through the k8s API. I'm not seeing value in making multiple levels of APIs to do that.

If we want to limit who or what can initiate reboot, having a dedicated CRD and using standard RBAC is the most straight-forward way to accomplish that. A webhook could potentially be used to restrict access to a specific field based on custom rules, but that's a much less obvious behavior. There's no way to programatically express to a client that its update will be rejected; it has to try the update and then deal with it being potentially rejected, and it won't know why. With RBAC, it's all clear up-front whether a client has the ability to update a resource, and it's very easy to give that permission.

Am I understanding correctly that whatever *implements* reboot will just do these things?
* set `Spec.Online = false`
* wait for `Status.PoweredOn == false`
* set `Spec.Online = true`

In the separate CRD case, the fencing code would just create an instance of that CRD instead of writing a timestamp to the BMH. Seems simple either way. Is there some aspect of correctness that would be sacrificed?

Reply:
> Am I understanding correctly that whatever implements reboot will just do these things?

Conceptually yes, but with the key constraint that `Spec.Online` can't actually change, because if it does then multiple reboot requests can interfere with each other and with unrelated power management updates from the user.

Say for example that the user decides they want to power down the host, and the request hits during a reboot. The reboot will turn the power back on at the end even though the user wanted the server to remain powered off, which is not correct. AFAIK this is not solvable by an external CRD without adding some other field in the BareMetalHost spec that is orthogonal to the `Online` field.

Reply:
Any time you have multiple actors that modify the spec of something, which is fairly common, those things can disagree about the desired state, and one of them will lose. It gets especially exciting if they're both controllers and they end up in a reconcile fight. We at least try to avoid that. :)

The best we can do is have a clear declarative API. If one actor requests a reboot at approximately the same time as another actor requests power off, we can just hope that whoever loses will keep watching and try again if appropriate.

Reply:
Right, but twiddling the Online field to effect a reboot is not a declarative API; it's the epitome of a workflow hacked in to a supposedly-declarative system (something I regrettably have a lot of experience with). Hence this patch, which adds a truly declarative API.